File: 369. Plus One Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def plusOne(self, head: ListNode) -> ListNode:
        
        def helper(node):
            if not node:
                return True
            add_one = helper(node.next)
            if add_one and node.val == 9:
                node.val = 0
                return True
            else:
                if add_one:
                    node.val += 1
                return False
        if helper(head):
            new_head = ListNode(1, head)
            return new_head
        else:
            return head

File: 369. Plus One Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode plusOne(ListNode head) {
        int add = helper(head);
        if(add == 1){
            ListNode new_head = new ListNode(1, head);
            return new_head;
        }else{
            return head;
        }
    }
    
    public int helper(ListNode node){
        if(node == null){
            return 1;
        }
        int add = helper(node.next);
        node.val += add;
        int ret = node.val / 10;
        node.val = node.val % 10;
        return ret;
    }
}